### ISSUE TITLE
unmatch of two ctc loss in pytorch backend

### ISSUE THREAD

**Issue by qmpzzpmq** [espnet/espnet#682]
https://github.com/espnet/espnet/blob/32dc93735d5e4d3e5b7edef743784d1582ecae85/espnet/nets/pytorch_backend/ctc.py#L29

when default ctc reduction is true, which should means that is reducted, and it should be "mean" here

**Comment by sw005320** [comment]
I just want to clarify. We intend not to perform "mean" across time but only across utterance. We keep not to normalize the cost function by the length. Does that "mean" works like that?

**Comment by qmpzzpmq** [comment]
@sw005320 here is some description of two ctc_loss about "mean" and "sum": https://discuss.pytorch.org/t/ctcloss-performance-of-pytorch-1-0-0/27524/6

let me clear that, if `reduce` is set true here, which means warp_ctc computation is normalized not only in utterances (by size_average=True) but also in time (by length_average = reduction). Meanwhile, the built-in ctc is set as "torch.nn.CTCLoss(reduction=sum)", which means the unnormalized value is return.  But if the `reduction=mean` here. the normalized both is return even the normalized method is different. 

You can find the explain in the link above.

**Comment by sw005320** [comment]
I see. This is not our intention. We'll fix it.

**Comment by qmpzzpmq** [comment]
if you want to normalize at time and utterance both, just change to "sum"

**Comment by qmpzzpmq** [comment]
Sorry  i mean “mean”

**Comment by kan-bayashi** [comment]
@qmpzzpmq @sw005320 Let me confirm. 
In the case of warp-ctc, we use `size_average=True`, so the normalization is performed in direction of batch, not length of each utterances.
https://github.com/jnishi/warp-ctc/blob/8002586417759c664ab056faf5d6d6a960ca8b6d/pytorch_binding/warpctc_pytorch/__init__.py#L35-L46

In the case of pytorch ctc, we use `reduction_type=“sum”`, so the normalization is not performed.
However, in `forward` we divide loss value by the number of batches.
https://github.com/espnet/espnet/blob/3bb0a25af88eedbd1456d253324199109f4a8203/espnet/nets/pytorch_backend/ctc.py#L41-L46

Therefore, the both normalization method should be the same.

**Comment by sw005320** [comment]
@kan-bayashi, I think you're right.
But I actually found that the CTC loss (warp CTC) is higher than (e.g., twice higher) before, and I was thinking that this may be a reason.
